llmfoundry/models/mpt/modeling_mpt.py

```diff
@@ -607,6 +607,13 @@ def forward(
             attn_uses_sequence_id=self.attn_uses_sequence_id,
             attn_impl=self.attn_impl,
             attention_mask=attention_mask)
+
+        if self.alibi and self.attn_impl == 'flash':
```

Comment:
how come attn_bias needs to be set for flash attention but not for other attention types like triton?

Reply:
Also I think this is missing the check for the flash attention version

Reply:
Regarding attn_bias for triton and torch: For those attention types, the attn_bias gets set [here](https://github.com/mosaicml/llm-foundry/blob/083b4b2451a60487f0ff891b4fbd671064b90a30/llmfoundry/models/layers/attention.py#L788). 

Regarding the missing check: I skipped the check because I added a check for the same in `configuration_mpt.py`
<img width="801" alt="Screenshot 2024-01-04 at 12 58 43 PM" src="https://github.com/mosaicml/llm-foundry/assets/144760128/7424662f-5922-4631-bd08-a44ebc213a5a">

Reply:
@ShashankMosaicML Can we set the attn_bias in the same way that it is set for triton and torch? We should unify these code paths if possible. 

I suggest adding that logic here: https://github.com/mosaicml/llm-foundry/blob/083b4b2451a60487f0ff891b4fbd671064b90a30/llmfoundry/models/layers/attention.py#L786 such that this function should handle alibi with fa2.

Additionally for my understanding, is there a reason why for fa2 we call gen_slopes directly instead of calling build_alibi_bias? Thanks!

Reply:
After discussion with @irenedea, we decided that since the attn_bias for triton and torch was different from what it was for flash attention, it would be better to have a separate variable for flash attention, which I named `alibi_slopes`. Please check the updated PR to see the changes.

FAv2 directly takes in the slopes corresponding to ALiBi's linear biases, whereas Triton and Torch use the actual biases. Hence for FAv2 we simply call gen_slopes instead of calling build_alibi_bias.